jepsen/test/jepsen/checker_test.clj

```diff
@@ -190,35 +191,79 @@
+                                                        8 25
+                                                        9 25
+                                                        10 25])))))
+
+(defn perf-gen
+  ([latency]
+   (perf-gen latency nil))
+  ([latency nemesis-regions]
+   (let [f (rand-nth [:write :read])
+         proc (rand-int 100)
+         time (* 1e9 (rand-int 100))
+         type (rand-nth [:ok :ok :ok :ok :ok
+                         :fail :info :info])]
+     [{:process proc, :type :invoke, :f f, :time time}
+      {:process proc, :type type,    :f f, :time
+       (+ time latency)}])))
 
 (deftest perf-test
-  (check (perf)
-         {:name       "perf test"
-          :start-time 0}
-         (->> (repeatedly #(/ 1e9 (inc (rand-int 1000))))
-              (mapcat (fn [latency]
-                        (let [f (rand-nth [:write :read])
-                              proc (rand-int 100)
-                              time (* 1e9 (rand-int 100))
-                              type (rand-nth [:ok :ok :ok :ok :ok
-                                              :fail :info :info])]
-                          [{:process proc, :type :invoke, :f f, :time time}
-                           {:process proc, :type type,    :f f, :time
-                            (+ time latency)}])))
-              (take 10000)
-              vec)
-         {}))
+  (let [history (->> (repeatedly #(/ 1e9 (inc (rand-int 1000))))
+                     (mapcat perf-gen)
+                     (take 10000)
+                     vec)]
+
+    (testing "can render latency-graph"
+      (is (= (check (latency-graph)
+                    {:name "latency graph"
+                     :start-time 0}
+                    history
+                    {})
+             {:valid? true})))
+
+    (testing "can render rate-graph"
+      (is (= (check (rate-graph)
+                    {:name "rate graph"
+                     :start-time 0}
+                    history
+                    {})
+             {:valid? true})))
+
+    (testing "can render combined perf graph"
+      (is (= (check (perf)
+                    {:name "perf graph"
+                     :start-time 0}
+                    history
+                    {})
+             {:latency-graph {:valid? true},
+              :rate-graph {:valid? true},
+              :valid? true})))
+
+    ;; TODO Generate nemesis regions
+    (testing "can accept nemesis regions"
+      (let [checker (perf {:nemeses #{{:start #{:start1}
+                                       :stop  #{:stop1}}
+                                      {:start #{:start2.1 :start2.2}
+                                       :stop  #{:stop2}}
+                                      {:start #{:start3}
+                                       :stop  #{:stop3.1 :stop3.2}}
+                                      {:start #{:start4.1 :start4.2 :start4.3}
+                                       :stop  #{:stop4.1 :stop4.2 :stop4.3}}}})
```

Comment:
This is the key section. Do you like this option structure? I'm preeeeeeetty sure I can get it working in the renderer, but I wanted to check in before digging in deeply on this approach.

Reply:
Yeah, I think that'll work just fine. :-)

Reply:
(and it makes sense to do maps with :start and :stop here in case we wanted to let you specify certain colors for nemeses, or maybe shading styles)

Reply:
:D Something just occurred to me: we could be better off using a vector of maps here instead of a set of maps. That way the top->down ordering of the nemesis bars/regions is explicit. Thoughts?

Reply:
Yeah, I'm OK with that too, though generally I have the plotting system sort things on display.